astropy/units/tests/test_format.py

```diff
@@ -500,15 +516,51 @@ def test_vounit_custom():
     assert x_string == 'm mfoo'
 
 
+def test_vounit_shadow():
+    """Test that a quoted custom unit 'shadowing' a physical unit is correctly
+    identified as 'unknown' and raises meaningful conversion errors.
+    """
+    with catch_warnings() as w:
+        x = u.Unit("'m'", format='vounit')
+        assert "Unit 'm' is not recognized in the VOUnit standard" in str(w[0])
+    assert u.physical.get_physical_type(x) == 'unknown'
+    with pytest.raises(u.UnitConversionError,
+                       match=r'\(custom unit ..m..\) and .cm. \(length\)'):
+        x.to('cm')
+
+
 def test_vounit_implicit_custom():
+    # We cannot overwrite already created custom units, so first make sure
+    # that the namespace is clean:
+    u.format.VOUnit._custom_units.clear()
```

Comment:
Not really necessary, but I think the proper way to write these types of test is to have a test class that has a `setup_class`/`teardown_class` method pair that ensures the directory stays in the same state.  Now, if someone runs `astropy.test()` their custom units disappear.

Reply:
Good point; lately I am noticing a bunch of problems from the opposite side with tests (particularly from `io.asdf` it seems) adding "enabled" units that later fail to raise expected warnings or exceptions.
Will have to look into this and the class definition issue above later when it's settled how or if to proceed with the PR.